### PR TITLE
API Keys: Add revocation for SATs

### PR DESCRIPTION
**Pull request by Jguer**
**What this PR does / why we need it**:

- Allow revoking of API keys (API endpoints not setup for now). This is a soft-delete, close to disable for users
- Rework SAT store to make more use the API Key service
- Move background service to run in the service and not the store

### CODE REVIEW

File: pkg/services/serviceaccounts/database/token_store_test.go [hunk 1]
```diff
@@ -91,6 +96,48 @@ func TestStore_AddServiceAccountToken_WrongServiceAccount(t *testing.T) {
...
+		Key:           key.HashedKey,
+		SecondsToLive: 0,
+		Result:        &apikey.APIKey{},
+	}
+
+	err = store.AddServiceAccountToken(context.Background(), sa.ID, &cmd)
+	require.NoError(t, err)
+	newKey := cmd.Result
+
+	// Delete key from wrong service account
```
Comment (Jguer) [inline pkg/services/serviceaccounts/database/token_store_test.go:120]:
```suggestion
	// Revoke SAT
```

File: pkg/services/sqlstore/migrations/apikey_mig.go [hunk 2]
```diff
@@ -95,4 +95,9 @@ func addApiKeyMigrations(mg *Migrator) {
 	mg.AddMigration("Add last_used_at to api_key table", NewAddColumnMigration(apiKeyV2, &Column{
 		Name: "last_used_at", Type: DB_DateTime, Nullable: true,
 	}))
+
+	// is_revoked indicates whether key is revoked or not. Revoked keys should be kept in the table, but invalid.
+	mg.AddMigration("Add is_disabled column to api_key table", NewAddColumnMigration(apiKeyV2, &Column{
```
Comment (Jguer) [inline pkg/services/sqlstore/migrations/apikey_mig.go:100]:
```suggestion
	mg.AddMigration("Add is_revoked column to api_key table", NewAddColumnMigration(apiKeyV2, &Column{
```

File: pkg/services/serviceaccounts/database/token_store.go [hunk 3]
```diff
@@ -2,27 +2,37 @@ package database
...
 import (
 	"context"
-	"time"
 
 	"github.com/grafana/grafana/pkg/services/apikey"
 	"github.com/grafana/grafana/pkg/services/org"
 	"github.com/grafana/grafana/pkg/services/serviceaccounts"
 	"github.com/grafana/grafana/pkg/services/sqlstore"
-	"xorm.io/xorm"
+	"github.com/pkg/errors"
```
Comment (kalleep) [inline pkg/services/serviceaccounts/database/token_store.go:10]: Nit: imports

File: pkg/services/serviceaccounts/database/token_store.go [hunk 4]
```diff
@@ -33,37 +43,27 @@ func (s *ServiceAccountsStoreImpl) AddServiceAccountToken(ctx context.Context, s
...
-			Updated:          updated,
-			Expires:          expires,
-			LastUsedAt:       nil,
-			ServiceAccountId: &serviceAccountId,
+			SecondsToLive:    cmd.SecondsToLive,
+			ServiceAccountID: &serviceAccountId,
 		}
 
-		if _, err := sess.Insert(&token); err != nil {
+		if err := s.apiKeyService.AddAPIKey(ctx, addKeyCmd); err != nil {
```
Comment (kalleep) [inline pkg/services/serviceaccounts/database/token_store.go:55]: We should call this from service account service instead of the store

**Review by kalleep — APPROVED** [review]
LGTM left some comments

File: pkg/services/apikey/model.go [hunk 5]
```diff
@@ -26,18 +26,21 @@ type APIKey struct {
 	LastUsedAt       *time.Time `xorm:"last_used_at"`
 	Expires          *int64
 	ServiceAccountId *int64
+	IsRevoked        *bool `xorm:"is_revoked"`
```
[the remark this answers is not in the document]
Comment (Jguer) [reply pkg/services/apikey/model.go:29]: I would keep it as the field is nullable in DB

File: pkg/services/serviceaccounts/api/token.go [hunk 6]
```diff
@@ -37,6 +37,8 @@ type TokenDTO struct {
 	SecondsUntilExpiration *float64 `json:"secondsUntilExpiration"`
 	// example: false
 	HasExpired bool `json:"hasExpired"`
+	// example: false
+	IsRevoked *bool `json:"isRevoked"`
```
[the remark this answers is not in the document]
Comment (Jguer) [reply pkg/services/serviceaccounts/api/token.go:41]: I would keep it as the field is nullable in DB

[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/serviceaccounts/api/token.go:41]: oh missed that part

**Pull request merged**